// play-dl/dist/YouTube/classes/Video.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.YouTubeVideo = void 0;
/**
 * Class for YouTube Video url
 */
class YouTubeVideo {
    constructor(data) {
        var _a, _b;
        if (!data)
            throw new Error(`Can not initiate ${this.constructor.name} without data`);
        this.id = data.id || undefined;
        this.url = `https://www.youtube.com/watch?v=${this.id}`;
        this.type = 'video';
        this.title = data.title || undefined;
        this.description = data.description || undefined;
        this.durationRaw = data.duration_raw || '0:00';
        this.durationInSec = (data.duration < 0 ? 0 : data.duration) || 0;
        this.uploadedAt = data.uploadedAt || undefined;
        this.views = parseInt(data.views) || 0;
        this.thumbnail = data.thumbnail || {};
        this.channel = data.channel || {};
        this.likes = ((_a = data.ratings) === null || _a === void 0 ? void 0 : _a.likes) || 0;
        this.dislikes = ((_b = data.ratings) === null || _b === void 0 ? void 0 : _b.dislikes) || 0;
        this.live = !!data.live;
        this.private = !!data.private;
        this.tags = data.tags || [];
    }
    get toString() {
        return this.url || '';
    }
    get toJSON() {
        var _a, _b, _c;
        return {
            id: this.id,
            url: this.url,
            title: this.title,
            description: this.description,
            durationInSec: this.durationInSec,
            durationRaw: this.durationRaw,
            uploadedAt: this.uploadedAt,
            thumbnail: this.thumbnail,
            channel: {
                name: (_a = this.channel) === null || _a === void 0 ? void 0 : _a.name,
                id: (_b = this.channel) === null || _b === void 0 ? void 0 : _b.id,
                icon: (_c = this.channel) === null || _c === void 0 ? void 0 : _c.iconURL()
            },
            views: this.views,
            type: this.type,
            tags: this.tags,
            ratings: {
                likes: this.likes,
                dislikes: this.dislikes
            },
            live: this.live,
            private: this.private
        };
    }
}
exports.YouTubeVideo = YouTubeVideo;
